### DISCUSSION TITLE
Cornell CS3110: OCaml Programming Correct + Efficient + Beautiful - CS自学指南

### DISCUSSION THREAD

**Discussion by giscus** [PKUFlyingPig/cs-self-learning#491] [Announcements]
# Cornell CS3110: OCaml Programming Correct + Efficient + Beautiful - CS自学指南

CS自学指南

https://csdiy.wiki/%E7%BC%96%E7%A8%8B%E5%85%A5%E9%97%A8/CS3110/

**Comment by PuJunhong** [comment]
学完后，还用学CS61A吗

**Reply by featherwit001** [reply]
SICP还是要学的，第一门编程课还是SCIP。
这个是函数式编程，内容量其实比SICP多很多，从语言本身讲到编程风格和数据结构，再到编译和特性实现，跨度非常大。
至于为什么要学函数式编程。举一个例子说明，红黑树如果用OCaml写150行绰绰有余，用C或者命令式风格400行都不够。函数式编程风格是一种全新的思考方式，会带来思维模型的彻底改变。

**Reply by PuJunhong** [reply]
谢谢大佬，学到了

**Comment by lingkerio** [comment]
看了一半左右OCaml的官网教程，今天准备系统学一下OCaml于是搜索了CS3110（之前就已经了解到这门课了），没想到csdiy恰好更新了这门课，感觉也算是缘分了，加油！

**Reply by featherwit001** [reply]
哈哈，这门课非常非常有意思，能扩展很多见识

**Reply by PuJunhong** [reply]
另外想问一下课程作业怎么选择？

**Reply by featherwit001** [reply]
二星题目，是用来复习引导的，如果一眼看过去就知道大概的程序框架，那做一两道就够了。
三星题目，主要是基本应用，如对视频中展示的代码都自己敲过，总结过，那么都是可以很容易做出来的。
四星题目，难度比较高，基本的流程是，先写docunment，对每个函数要实现什么功能进行描述，然后写测试，最后写实现，这个是耗时最短的流程。
课程作业，凭自己的兴趣，选择几道三星的题目，提高一些熟练度，然后多做4星的题目。如果发现几个题目是一个系列（名称类似），那么也可以做。

**Comment by featherwit001** [comment]
另外补充一点，这门课程串联了很多东西，可以尽自己所能去探索实现某一个方面的内容，
比如：课中详细介绍的函数式三大操作，map，fold，filter具体的实现细节，尽可能地熟悉和使用

比如尾递归优化，参考课程教材，对所有之前设计到递归的函数改写成尾递归优化的形式，哪些能改，那么不能改？

比如：作业中的矩阵，尝试实现矩阵运算，加减乘除，为了加快矩阵乘法可以尝试使用矩阵转置，使用QCheck测试框架，看看优化前后面对不同规模的数据（100维至2000维）实际的耗时有什么变化？

比如：使用AVL tree实现Map，可以参考OCaml Stdlib中的实现，实现中平衡因子是2，而是不是标准的1，思考这是为什么？

比如：课程中只涉及了红黑树的插入平衡，可以探索红黑树的插入，使用红黑树重新实现Map。

比如： 参照课程提供的calc，实现完成的包含浮点数的四则运算，注意负数怎么实现？

比如：参照课程的simPL语言，探索一下函数fun和apply 如何实现？

比如：参照课程的simPL语言，探索一下函数fun和apply 如何实现？类型检查type check 如何实现

比如：参照课程介绍的 Hindley-Milner algorithm，类型推断type reference 的原理是什么？其他语言，如Haskell，rust，racket 都使用了什么算法实现类型推断？

这些是我做过的一部分，还有更多可以探索的东西，而且这些问题的起点都是课程中的内容，从课中出发，按照自己的兴趣，使用所搜索和chatgpt，能充分运用学到的知识。

**Comment by HarryLu0708** [comment]
下学期学ocaml，先学上

**Comment by hiyasame** [comment]
请问这门课可以代替sicp吗

**Reply by featherwit001** [reply]
两门课重合的地方很少

**Comment by acheac** [comment]
学完后脑袋空空，考试代码打的我发冷汗😭

**Reply by lingkerio** [reply]
佬在Cornell就读吗？考试内容有没有兴趣和闲暇分享一下下（（（（逃

**Reply by acheac** [reply]
不在cornell，学校刚好也有这门课 （

**Comment by Gravifer** [comment]
我会推荐先学haskell mooc再来做这门课； haskell mooc的练习更精简并且更hands on一些，相比起本书的1,2星练习可能更能形成反馈

**Comment by Garhlz** [comment]
教材的中文翻译，机翻+ai润色：https://garhlz.github.io/cs3110-textbook/
